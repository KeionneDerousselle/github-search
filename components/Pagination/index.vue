<template>
  <nav aria-label="Search results navigation">
    <ul class="pagination">
      <li class="page__item page__item--previous">
        <a
          href="#"
          :class="['page__link', {'page__link--disabled': firstPageIsActive}]"
          :disabled="firstPageIsActive"
          @click="handlePreviousClicked">
          <chevron-left-icon
            aria-hidden="true"
            class="page__link__icon" />

          <span class="sr-only">Navigate to Previous Page</span>
        </a>
      </li>

      <li
        v-for="page in pages"
        :key="page.label"
        :class="['page__item', 'page__item--page', `page__item--${page.label}`]"
        :aria-label="`Navigate to page number ${page.label}`">
        <a
          href="#"
          :class="['page__link', `page__link--${page.label}`, {'page__link--active': page.active} ]"
          :disabled="page.active"
          @click="handlePageClicked(page)">
          {{ page.label }}
        </a>
      </li>

      <li class="page__item page__item--next">
        <a
          href="#"
          :class="['page__link', {'page__link--disabled': lastPageIsActive }]"
          :disabled="lastPageIsActive"
          @click="handleNextClicked">
          <chevron-right-icon
            aria-hidden="true"
            class="page__link__icon" />

          <span class="sr-only">Navigate to Next Page</span>
        </a>
      </li>
    </ul>
  </nav>
</template>
<script>
import ChevronLeftIcon from '@/components/Icons/ChevronLeft'
import ChevronRightIcon from '@/components/Icons/ChevronRight'

export default {
  components: {
    ChevronLeftIcon,
    ChevronRightIcon
  },

  props: {
    currentPage: {
      type: Number,
      default: 1
    },

    itemsPerPage: {
      type: Number,
      required: true
    },

    totalItems: {
      type: Number,
      required: true
    },

    maxVisibilePages: {
      type: Number,
      default: 3
    }
  },

  computed: {
    firstVisiblePage() {
      if (this.currentPage === 1) {
        return 1
      } else if (this.currentPage === this.totalPages) {
        return this.totalPages - this.maxVisibilePages + 1
      } else {
        return this.currentPage - 1
      }
    },

    totalPages() {
      return Math.ceil(this.totalItems / this.itemsPerPage)
    },

    lastVisiblePage() {
      return Math.min(this.firstVisiblePage + this.maxVisibilePages - 1, this.totalPages)
    },

    pages() {
      const currentPages = []

      for (let i = this.firstVisiblePage; i <= this.lastVisiblePage; i++) {
        currentPages.push({
          label: i,
          active: i === this.currentPage
        })
      }

      return currentPages
    },

    firstPageIsActive() {
      return this.currentPage === 1
    },

    lastPageIsActive() {
      return this.currentPage === this.totalPages
    }
  },

  methods: {
    handlePageClicked(page) {
      if (!page.active) {
        this.$emit('change', page.label)
      }
    },

    handleNextClicked() {
      if (this.currentPage < this.totalPages) {
        this.$emit('change', this.currentPage + 1)
      }
    },

    handlePreviousClicked() {
      if (this.currentPage > 1) {
        this.$emit('change', this.currentPage - 1)
      }
    }
  }
}
</script>
<style lang="scss">
.pagination {
  @apply flex w-full h-full clear-both;

  .page__item {
    &--page:not(:last-of-type) {
      @apply ml-2;
    }

    &--previous {
      @apply mr-6;
    }

    &--next {
      @apply ml-6;
    }
  }

  .page__link {
    @apply rounded-full border border-transparent inline-flex w-10 h-10 flex items-center justify-center text-sm md:text-base;

    transition: all 0.3s ease-in;

    &:hover:not(.page__link--disabled),
    &:focus:not(.page__link--disabled) {
      @apply border-indigo-100;
    }

    &__icon {
      @apply text-indigo-100 w-4 h-4;
    }

    &--active {
      @apply bg-indigo-200 text-indigo-900;
    }

    &--disabled {
      @apply cursor-not-allowed bg-gray-300 bg-opacity-60;

      .page__link__icon {
        @apply text-gray-600 w-4 h-4;
      }
    }
  }
}
</style>
