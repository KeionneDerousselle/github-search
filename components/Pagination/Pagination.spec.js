import Pagination from '.'

describe('Pagination', () => {
  let wrapper

  describe('when the first page is active', () => {
    beforeAll(async () => {
      wrapper = shallowPreMocked(Pagination, {
        propsData: {
          itemsPerPage: 25,
          totalItems: 75
        }
      })

      await wrapper.vm.$nextTick()
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should disable the previous link', () => {
      const prev = wrapper.get('.page__item--previous .page__link')

      expect(prev.element).toBeVisible()
      expect(prev.attributes().disabled).toBeTruthy()
      expect(prev.classes()).toContain('page__link--disabled')
    })

    it('should enable the next page link', () => {
      const next = wrapper.get('.page__item--next .page__link')

      expect(next.element).toBeVisible()
      expect(next.attributes().disabled).toBeFalsy()
      expect(next.classes()).not.toContain('page__link--disabled')
    })

    it('should set the current page as the active page', () => {
      const currentPage = wrapper.get(`.page__link--${wrapper.vm.currentPage}`)

      expect(currentPage.element).toBeVisible()
      expect(currentPage.attributes().disabled).toBeTruthy()
      expect(currentPage.classes()).toContain('page__link--active')
    })

    it('should enable all other pages', () => {
      wrapper.vm.pages.forEach(page => {
        if (!page.active) {
          const pageEl = wrapper.get(`.page__link--${page.label}`)

          expect(pageEl.element).toBeVisible()
          expect(pageEl.attributes().disabled).toBeFalsy()
          expect(pageEl.classes()).not.toContain('page__link--active')
        }
      })
    })
  })

  describe('when the first page is not active', () => {
    beforeAll(() => {
      wrapper = shallowPreMocked(Pagination, {
        propsData: {
          itemsPerPage: 25,
          totalItems: 75,
          currentPage: 2
        }
      })
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should enable the previous link', () => {
      const prev = wrapper.get('.page__item--previous .page__link')

      expect(prev.element).toBeVisible()
      expect(prev.attributes().disabled).toBeFalsy()
      expect(prev.classes()).not.toContain('page__link--disabled')
    })

    it('should enable the next page link', () => {
      const next = wrapper.get('.page__item--next .page__link')

      expect(next.element).toBeVisible()
      expect(next.attributes().disabled).toBeFalsy()
      expect(next.classes()).not.toContain('page__link--disabled')
    })

    it('should set the current page as the active page', () => {
      const currentPage = wrapper.get(`.page__link--${wrapper.vm.currentPage}`)

      expect(currentPage.element).toBeVisible()
      expect(currentPage.attributes().disabled).toBeTruthy()
      expect(currentPage.classes()).toContain('page__link--active')
    })

    it('should enable all other pages', () => {
      wrapper.vm.pages.forEach(page => {
        if (!page.active) {
          const pageEl = wrapper.get(`.page__link--${page.label}`)

          expect(pageEl.element).toBeVisible()
          expect(pageEl.attributes().disabled).toBeFalsy()
          expect(pageEl.classes()).not.toContain('page__link--active')
        }
      })
    })
  })

  describe('when the last page is active', () => {
    beforeAll(() => {
      wrapper = shallowPreMocked(Pagination, {
        propsData: {
          itemsPerPage: 25,
          totalItems: 75,
          currentPage: 3
        }
      })
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should enable the previous link', () => {
      const prev = wrapper.get('.page__item--previous .page__link')

      expect(prev.element).toBeVisible()
      expect(prev.attributes().disabled).toBeFalsy()
      expect(prev.classes()).not.toContain('page__link--disabled')
    })

    it('should disable the next page link', () => {
      const next = wrapper.get('.page__item--next .page__link')

      expect(next.element).toBeVisible()
      expect(next.attributes().disabled).toBeTruthy()
      expect(next.classes()).toContain('page__link--disabled')
    })

    it('should set the current page as the active page', () => {
      const currentPage = wrapper.get(`.page__link--${wrapper.vm.currentPage}`)

      expect(currentPage.element).toBeVisible()
      expect(currentPage.attributes().disabled).toBeTruthy()
      expect(currentPage.classes()).toContain('page__link--active')
    })

    it('should enable all other pages', () => {
      wrapper.vm.pages.forEach(page => {
        if (!page.active) {
          const pageEl = wrapper.get(`.page__link--${page.label}`)

          expect(pageEl.element).toBeVisible()
          expect(pageEl.attributes().disabled).toBeFalsy()
          expect(pageEl.classes()).not.toContain('page__link--active')
        }
      })
    })
  })

  describe('when the last page is not active', () => {
    beforeAll(() => {
      wrapper = shallowPreMocked(Pagination, {
        propsData: {
          itemsPerPage: 25,
          totalItems: 75,
          currentPage: 2
        }
      })
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should enable the previous link', () => {
      const prev = wrapper.get('.page__item--previous .page__link')

      expect(prev.element).toBeVisible()
      expect(prev.attributes().disabled).toBeFalsy()
      expect(prev.classes()).not.toContain('page__link--disabled')
    })

    it('should enable the next page link', () => {
      const next = wrapper.get('.page__item--next .page__link')

      expect(next.element).toBeVisible()
      expect(next.attributes().disabled).toBeFalsy()
      expect(next.classes()).not.toContain('page__link--disabled')
    })

    it('should set the current page as the active page', () => {
      const currentPage = wrapper.get(`.page__link--${wrapper.vm.currentPage}`)

      expect(currentPage.element).toBeVisible()
      expect(currentPage.attributes().disabled).toBeTruthy()
      expect(currentPage.classes()).toContain('page__link--active')
    })

    it('should enable all other pages', () => {
      wrapper.vm.pages.forEach(page => {
        if (!page.active) {
          const pageEl = wrapper.get(`.page__link--${page.label}`)

          expect(pageEl.element).toBeVisible()
          expect(pageEl.attributes().disabled).toBeFalsy()
          expect(pageEl.classes()).not.toContain('page__link--active')
        }
      })
    })
  })

  describe('when clicking a page that is active', () => {
    beforeAll(async () => {
      wrapper = shallowPreMocked(Pagination, {
        propsData: {
          itemsPerPage: 25,
          totalItems: 75,
          currentPage: 2
        }
      })

      const currentPage = wrapper.get(`.page__link--${wrapper.vm.currentPage}`)

      currentPage.trigger('click')

      await wrapper.vm.$nextTick()
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should not emit the change page event', () => {
      expect(wrapper.emitted().change).toBeFalsy()
    })
  })

  describe('when clicking a page that is not active', () => {
    beforeAll(async () => {
      wrapper = shallowPreMocked(Pagination, {
        propsData: {
          itemsPerPage: 25,
          totalItems: 75,
          currentPage: 2
        }
      })

      const page = wrapper.get('.page__link--1')

      page.trigger('click')

      await wrapper.vm.$nextTick()
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should emit the change page event with the clicked page', () => {
      expect(wrapper.emitted().change[0]).toEqual([1])
    })
  })

  describe('when clicking a next page link while the last page is active', () => {
    beforeAll(async () => {
      wrapper = shallowPreMocked(Pagination, {
        propsData: {
          itemsPerPage: 25,
          totalItems: 75,
          currentPage: 3
        }
      })

      const nextPageLink = wrapper.get('.page__item--next .page__link')

      nextPageLink.trigger('click')

      await wrapper.vm.$nextTick()
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should not emit the change page event', () => {
      expect(wrapper.emitted().change).toBeFalsy()
    })
  })

  describe('when clicking a next page link while the last page is not active', () => {
    beforeAll(async () => {
      wrapper = shallowPreMocked(Pagination, {
        propsData: {
          itemsPerPage: 25,
          totalItems: 75,
          currentPage: 2
        }
      })

      const nextPageLink = wrapper.get('.page__item--next .page__link')

      nextPageLink.trigger('click')

      await wrapper.vm.$nextTick()
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should emit the change page event with the next page', () => {
      expect(wrapper.emitted().change[0]).toEqual([3])
    })
  })

  describe('when clicking a previous page link while the first page is active', () => {
    beforeAll(async () => {
      wrapper = shallowPreMocked(Pagination, {
        propsData: {
          itemsPerPage: 25,
          totalItems: 75,
          currentPage: 1
        }
      })

      const previousPageLink = wrapper.get('.page__item--previous .page__link')

      previousPageLink.trigger('click')

      await wrapper.vm.$nextTick()
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should not emit the change page event', () => {
      expect(wrapper.emitted().change).toBeFalsy()
    })
  })

  describe('when clicking a previous page link while the first page is not active', () => {
    beforeAll(async () => {
      wrapper = shallowPreMocked(Pagination, {
        propsData: {
          itemsPerPage: 25,
          totalItems: 75,
          currentPage: 2
        }
      })

      const previousPageLink = wrapper.get('.page__item--previous .page__link')

      previousPageLink.trigger('click')

      await wrapper.vm.$nextTick()
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should emit the change page event with the previous page', () => {
      expect(wrapper.emitted().change[0]).toEqual([1])
    })
  })
})
