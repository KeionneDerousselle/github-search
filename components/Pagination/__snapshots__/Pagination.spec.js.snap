// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination when the first page is active should render as expected 1`] = `
<nav aria-label="Search results navigation">
  <ul class="pagination">
    <li class="page__item page__item--previous"><a href="#" disabled="disabled" class="page__link page__link--disabled">
        <chevron-left-icon-stub aria-hidden="true" class="page__link__icon"></chevron-left-icon-stub> <span class="sr-only">Navigate to Previous Page</span>
      </a></li>
    <li aria-label="Navigate to page number 1" class="page__item page__item--page page__item--1"><a href="#" disabled="disabled" class="page__link page__link--1 page__link--active">
        1
      </a></li>
    <li aria-label="Navigate to page number 2" class="page__item page__item--page page__item--2"><a href="#" class="page__link page__link--2">
        2
      </a></li>
    <li aria-label="Navigate to page number 3" class="page__item page__item--page page__item--3"><a href="#" class="page__link page__link--3">
        3
      </a></li>
    <li class="page__item page__item--next"><a href="#" class="page__link">
        <chevron-right-icon-stub aria-hidden="true" class="page__link__icon"></chevron-right-icon-stub> <span class="sr-only">Navigate to Next Page</span>
      </a></li>
  </ul>
</nav>
`;

exports[`Pagination when the first page is not active should render as expected 1`] = `
<nav aria-label="Search results navigation">
  <ul class="pagination">
    <li class="page__item page__item--previous"><a href="#" class="page__link">
        <chevron-left-icon-stub aria-hidden="true" class="page__link__icon"></chevron-left-icon-stub> <span class="sr-only">Navigate to Previous Page</span>
      </a></li>
    <li aria-label="Navigate to page number 1" class="page__item page__item--page page__item--1"><a href="#" class="page__link page__link--1">
        1
      </a></li>
    <li aria-label="Navigate to page number 2" class="page__item page__item--page page__item--2"><a href="#" disabled="disabled" class="page__link page__link--2 page__link--active">
        2
      </a></li>
    <li aria-label="Navigate to page number 3" class="page__item page__item--page page__item--3"><a href="#" class="page__link page__link--3">
        3
      </a></li>
    <li class="page__item page__item--next"><a href="#" class="page__link">
        <chevron-right-icon-stub aria-hidden="true" class="page__link__icon"></chevron-right-icon-stub> <span class="sr-only">Navigate to Next Page</span>
      </a></li>
  </ul>
</nav>
`;

exports[`Pagination when the last page is active should render as expected 1`] = `
<nav aria-label="Search results navigation">
  <ul class="pagination">
    <li class="page__item page__item--previous"><a href="#" class="page__link">
        <chevron-left-icon-stub aria-hidden="true" class="page__link__icon"></chevron-left-icon-stub> <span class="sr-only">Navigate to Previous Page</span>
      </a></li>
    <li aria-label="Navigate to page number 1" class="page__item page__item--page page__item--1"><a href="#" class="page__link page__link--1">
        1
      </a></li>
    <li aria-label="Navigate to page number 2" class="page__item page__item--page page__item--2"><a href="#" class="page__link page__link--2">
        2
      </a></li>
    <li aria-label="Navigate to page number 3" class="page__item page__item--page page__item--3"><a href="#" disabled="disabled" class="page__link page__link--3 page__link--active">
        3
      </a></li>
    <li class="page__item page__item--next"><a href="#" disabled="disabled" class="page__link page__link--disabled">
        <chevron-right-icon-stub aria-hidden="true" class="page__link__icon"></chevron-right-icon-stub> <span class="sr-only">Navigate to Next Page</span>
      </a></li>
  </ul>
</nav>
`;

exports[`Pagination when the last page is not active should render as expected 1`] = `
<nav aria-label="Search results navigation">
  <ul class="pagination">
    <li class="page__item page__item--previous"><a href="#" class="page__link">
        <chevron-left-icon-stub aria-hidden="true" class="page__link__icon"></chevron-left-icon-stub> <span class="sr-only">Navigate to Previous Page</span>
      </a></li>
    <li aria-label="Navigate to page number 1" class="page__item page__item--page page__item--1"><a href="#" class="page__link page__link--1">
        1
      </a></li>
    <li aria-label="Navigate to page number 2" class="page__item page__item--page page__item--2"><a href="#" disabled="disabled" class="page__link page__link--2 page__link--active">
        2
      </a></li>
    <li aria-label="Navigate to page number 3" class="page__item page__item--page page__item--3"><a href="#" class="page__link page__link--3">
        3
      </a></li>
    <li class="page__item page__item--next"><a href="#" class="page__link">
        <chevron-right-icon-stub aria-hidden="true" class="page__link__icon"></chevron-right-icon-stub> <span class="sr-only">Navigate to Next Page</span>
      </a></li>
  </ul>
</nav>
`;
